src/field/gen_fieldops.py

```diff
@@ -0,0 +1,338 @@
+    # dimension individually, or the total number of elements
+    if elementwise:
+        low_level_length_arg = ",".join(["int n{}".format(d) for d in out.dimensions])
+        dims = {"n" + x: x for x in out.dimensions}
+    else:
+        low_level_length_arg = " int len"
+        dims = {"len": 'i'}
+
+    # This produces either a triple-nested loop, or a single loop over
+    # all elements (if elementwise is False)
+    for_loop = ""
+    for dimension, index in dims.items():
+        for_loop += for_loop_statement_template.format(dimension=dimension,
+                                                       index=index)
+
+    # Check that the mesh is the same for all fields
+    mesh_equality_assert = ""
+    compound_mesh_equality_assert = ""
+    if lhs != 'BoutReal' and rhs != 'BoutReal':
+        mesh_equality_assert = "ASSERT1(localmesh == rhs.getMesh());"
+        compound_mesh_equality_assert = "ASSERT1(fieldmesh == rhs.getMesh());"
+
+    # Either total number of elements, or size of each dimension separately
+    m = ',' if elementwise else '*'
+    non_compound_length_arg = m.join(["localmesh->LocalN{}".format(d) for d in out.dimensions])
+    compound_length_arg = m.join(["fieldmesh->LocalN{}".format(d) for d in out.dimensions])
+
+    # Only check field location for Field3D
+    non_compound_location_check = ""
+    compound_location_check = ""
+    if lhs == rhs == 'Field3D':
+        non_compound_location_check = location_check_template.format(operator_name=operator_name)
+        compound_location_check = compound_location_check_template.format(
+            operator_name=operator_name)
+
+    # Set location for result (again, only for Field3d)
+    non_compound_location_set = ""
+    if out == 'Field3D':
+        source = "rhs" if rhs == "Field3D" else "lhs"
+        non_compound_location_set = location_set_template.format(location_source=source)
+
+    mesh_source = "lhs" if lhs != 'BoutReal' else "rhs"
+
+    template_args = {
+        'out': out,
+        'lhs': lhs,
+        'rhs': rhs,
+
+        'operator': operator,
+        'operator_name': operator_name,
+
+        'low_level_length_arg': low_level_length_arg,
+        'non_compound_length_arg': non_compound_length_arg,
+        'compound_length_arg': compound_length_arg,
+
+        'for_loop': for_loop,
+
+        'mesh_equality_assert': mesh_equality_assert,
+        'compound_mesh_equality_assert': compound_mesh_equality_assert,
+
+        'mesh_source': mesh_source,
+        'non_compound_location_check': non_compound_location_check,
+        'non_compound_location_set': non_compound_location_set,
+        'compound_location_check': compound_location_check,
+
+        'result_op': out.getElement(data=elementwise),
+        'lhs_op': lhs.getElement(data=elementwise),
+        'rhs_op': rhs.getElement(data=elementwise),
+    }
+
+    return template_args
+
+
+if __name__ == "__main__":
+    for_gcc = True
```

Comment:
note that this might lead to compatibility issues.
If I remember correctly this has lead to a significant speed-up.
Should we aim to detect this at compile time in C++ using macros?
Otherwise, could we compile this using a C99 compiler, and use restrict? This would probably be the clean solution, and a C99 compiler should be basically always available as well ...

Reply:
We've done quite a lot of testing on various looping schemes, and the fastest in the majority (all?) cases is the "consecutive-blocked region-based single data iterator" scheme, beating even C-loops on `__restrict__` pointers. I think that scheme is about ready to go in, so we'll be switching to that soon anyway.